rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,1060 @@
+    rcutils_reset_error();
+    RMW_SET_ERROR_MSG(error_msg.str);
+    goto fail;
+  }
+  if (security_contexts) {
+    rcutils_ret =
+      rcutils_string_array_init(security_contexts, nodes_number, allocator);
+    if (RCUTILS_RET_OK != rcutils_ret) {
+      rcutils_error_string_t error_msg = rcutils_get_error_string();
+      rcutils_reset_error();
+      RMW_SET_ERROR_MSG(error_msg.str);
+      goto fail;
+    }
+  }
+  {
+    size_t j = 0;
+    for (const auto & elem : participants_) {
+      const auto & nodes_info = elem.second;
+      for (const auto & node_info : nodes_info.node_entities_info_seq) {
+        node_names->data[j] = rcutils_strdup(node_info.node_name.c_str(), *allocator);
+        if (!node_names->data[j]) {
+          goto fail;
+        }
+        node_namespaces->data[j] = rcutils_strdup(
+          node_info.node_namespace.c_str(), *allocator);
+        if (!node_namespaces->data[j]) {
+          goto fail;
+        }
+        if (security_contexts) {
+          security_contexts->data[j] = rcutils_strdup(
+            nodes_info.security_context.c_str(), *allocator);
+          if (!security_contexts->data[j]) {
+            goto fail;
+          }
+        }
+        j++;
+      }
+    }
+  }
+  return RMW_RET_OK;
+
+fail:
+  rcutils_ret = rcutils_string_array_fini(node_names);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rmw_dds_common",
+      "failed to cleanup during error handling: %s", rcutils_get_error_string().str);
+  }
+  rcutils_ret = rcutils_string_array_fini(node_namespaces);
+  if (rcutils_ret != RCUTILS_RET_OK) {
+    RCUTILS_LOG_ERROR_NAMED(
+      "rmw_dds_common",
+      "failed to cleanup during error handling: %s", rcutils_get_error_string().str);
+  }
+  return RMW_RET_BAD_ALLOC;
+}
+
+std::ostream &
+rmw_dds_common::operator<<(std::ostream & ostream, const GraphCache & graph_cache)
+{
+  std::lock_guard<std::mutex> guard(graph_cache.mutex_);
+  std::ostringstream ss;
+
+  ss << "---------------------------------" << std::endl;
+  ss << "Graph cache:" << std::endl;
+  ss << "  Discovered data writers:" << std::endl;
+  for (const auto & data_writer_pair : graph_cache.data_writers_) {
+    ss << "    gid: '" << data_writer_pair.first << "', topic name: '" <<
+      data_writer_pair.second.topic_name << "', topic_type: '" <<
+      data_writer_pair.second.topic_type << "'" << std::endl;
+  }
+  ss << "  Discovered data readers:" << std::endl;
+  for (const auto & data_reader_pair : graph_cache.data_readers_) {
+    ss << "    gid: '" << data_reader_pair.first << "', topic name: '" <<
+      data_reader_pair.second.topic_name << "', topic_type: '" <<
+      data_reader_pair.second.topic_type << "'" << std::endl;
+  }
+  ss << "  Discovered participants:" << std::endl;
+  for (const auto & item : graph_cache.participants_) {
+    ss << "    gid: '" << item.first << std::endl;
+    ss << "    context name '" << item.second.security_context << std::endl;
```

Comment:
@ivanpauno hmm, 
```suggestion
    ss << "    security context name '" << item.second.security_context << std::endl;
```
?

Otherwise it may lead to confusion.

Reply:
Done in https://github.com/ros2/rmw_dds_common/pull/4/commits/7579823e226e55aaac812b2fa9182dbe1fec5472.